blog/build-a-network-speed-monitor-in-bash-with-awk.md

```diff
@@ -0,0 +1,110 @@
+# Build a network speed monitor in bash with AWK
+
+Sometimes we need to measure the network speed of a server but some services like `speedtest` are not useful because we want to check our bandwidth with another cloud service and not the internet itself.
+
+## AWK to the rescue
+
+We can write a simple script in AWK to parse the `/proc/net/dev` pseudo-file which contains network device status information. AWK is a [programming language](https://dev.to/rrampage/awk---a-useful-little-language-2fhf) for manipulating columns of data that we can't solve with a simple `sed` command and a regex.
+
+But before talk about the AWK script, we should define the sentences in bash that we need in this script. Or, you can jump to the end to find the complete script 😁.
+
+## The bash part
+
+We need to create a `.sh` file with an infinite loop to print the network speed every 2 seconds.
+
+```bash
+INTERFACE=eth0
+echo "$INTERFACE  down (KiB/s)   up (KiB/s)"
+while :
+do
+  # The AWK command here!
+  sleep 2;
+done
+```
+
+First, we save the name of network interface to inspect in a `INTERFACE` variable, in this example is `eth0`.
+
+Second, we print a nice header to remember us which column is the download speed (and the upload one) and the speeds are measured in kilobytes per second (KiB/s, the `i` is because [1 Kibibyte is 1024 bytes](https://en.wikipedia.org/wiki/Byte#Units_based_on_powers_of_2)).
+
+Then, we start the infinite loop with a sleep of 2 seconds between every iteration. Don't worry, this script will be stopped using `Ctrl+C`.
+
+### Comparing the dev pseudo-file in time
+
+We want to measure the network speed in KiB per second, so we should read the `/proc/net/dev` file and compare with a new read from the dev file after 1 second. For this we can use a little bash trick called [process substitution](https://tldp.org/LDP/abs/html/process-sub.html).
+
+```bash
+awk '...' \
+  <(grep $INTERFACE /proc/net/dev) \
+  <(sleep 1; grep $INTERFACE /proc/net/dev)
+```
+
+We use `grep` to read the line of our interface, and after wait for 1 second we must read the line of our interface again.
+
+![Mr. Bean magic](https://dev-to-uploads.s3.amazonaws.com/i/mmadhlrajvdl2ed1m64l.jpg)
+
+## The AWK script
+
+If we check the output of `/proc/net/dev`
+
+```bash
+cat /proc/net/dev
+```
+
+We'll see that the 1st column is the interface name, the 2nd column is the received bytes (`rx`) and the 10th column is the transmitted bytes (`tx`).
+
+We should save the first read in the `rx` and `tx` variables, and then `print` the difference of the 2nd read with the first one. But, we could divide the result by 1024 because KiB/s is a better human readable format than B/s.
```

Comment:
Shouldn't this be 'the difference between the 2nd read and the first one'?

Reply:
Use `1st` + `2nd`, or `first` + `second`